File: src/Samples/FSharp/MinimalZeroInstall.fs
open NanoByte.Common.Tasks
open ZeroInstall.Services
open ZeroInstall.Store.Model

let services = new ServiceLocator(new CliTaskHandler())
let solve = services.Solver.Solve
let uncached = services.SelectionsManager.GetUncachedImplementations
let fetch = services.Fetcher.Fetch
let execute = services.Executor.Start

let run requirements =
    let selections = solve requirements
    fetch (uncached selections)
    execute selections

[<EntryPoint>]
let main args =
    ignore(run(new Requirements(args.[0])))
    0

File: src/Samples/CSharp/MinimalZeroInstall.cs
using NanoByte.Common.Tasks;
using ZeroInstall.Services;
using ZeroInstall.Store.Model;

class MinimalZeroInstall : ServiceLocator
{
    public static void Main(string[] args)
    {
        new MinimalZeroInstall().Run(new Requirements(args[0]));
    }

    public MinimalZeroInstall() : base(new CliTaskHandler())
    {}

    private void Run(Requirements requirements)
    {
        var selections = Solver.Solve(requirements);
        var missing = SelectionsManager.GetUncachedImplementations(selections);
        Fetcher.Fetch(missing);
        Executor.Start(selections);
    }
}
